src/Components/Endpoints/src/Binding/DefaultFormValuesSupplier.cs

```diff
@@ -1,82 +1,113 @@
-
     private readonly HttpContextFormDataProvider _formData;
     private readonly FormDataMapperOptions _options = new();
-    private static readonly ConcurrentDictionary<Type, Func<IReadOnlyDictionary<string, StringValues>, FormDataMapperOptions, string, object>> _cache =
-        new();
+    private static readonly ConcurrentDictionary<Type, FormValueSupplier> _cache = new();
 
     public DefaultFormValuesSupplier(FormDataProvider formData)
     {
         _formData = (HttpContextFormDataProvider)formData;
     }
 
-    public bool CanBind(string formName, Type valueType)
-    {
-        return _formData.IsFormDataAvailable &&
-            string.Equals(formName, _formData.Name, StringComparison.Ordinal) &&
-            _options.ResolveConverter(valueType) != null;
-    }
-
-    public bool TryBind(string formName, Type valueType, [NotNullWhen(true)] out object? boundValue)
+    public bool CanBind(Type valueType, string? formName = null)
     {
-        // This will func to a proper binder
-        if (!CanBind(formName, valueType))
+        if (formName == null)
         {
-            boundValue = null;
-            return false;
+            return _options.ResolveConverter(valueType) != null;
         }
+        else
+        {
+            var result = _formData.IsFormDataAvailable &&
+                string.Equals(formName, _formData.Name, StringComparison.Ordinal) &&
+                _options.ResolveConverter(valueType) != null;
 
-        var deserializer = _cache.GetOrAdd(valueType, CreateDeserializer);
+            return result;
+        }
+    }
 
-        var result = deserializer(_formData.Entries, _options, "value");
-        if (result != default)
+    public void Bind(FormValueSupplierContext context)
+    {
+        // This will func to a proper binder
+        if (!CanBind(context.ValueType, context.FormName))
         {
-            // This is not correct, but works for primtive values.
-            // Will change the interface when we add support for complex types.
-            boundValue = result;
-            return true;
+            context.SetResult(null);
         }
 
-        boundValue = valueType.IsValueType ? Activator.CreateInstance(valueType) : null;
-        return false;
+        var deserializer = _cache.GetOrAdd(context.ValueType, CreateDeserializer);
+        Debug.Assert(deserializer != null);
+        deserializer.Deserialize(context, _options, _formData.Entries, _formData.FormOptions);
     }
 
-    private Func<IReadOnlyDictionary<string, StringValues>, FormDataMapperOptions, string, object> CreateDeserializer(Type type) =>
-        _method.MakeGenericMethod(type)
-        .CreateDelegate<Func<IReadOnlyDictionary<string, StringValues>, FormDataMapperOptions, string, object>>();
+    private FormValueSupplier CreateDeserializer(Type type) =>
+        (FormValueSupplier)Activator.CreateInstance(typeof(FormValueSupplier<>)
+        .MakeGenericType(type))!;
```

Comment:
Is there anything we should be concerned about regarding AOT/trimming when using `MakeGenericType`? I know that the rest of the server-side system likely isn't fully compatible with AOT/trimming yet so it's not like this would be a regression. Just thinking ahead to if/when we try to get full AOT/trim compat. Will this be a blocker, or does it just work, or will there be some other way to achieve what we want at that point?

I know that source generators can in some cases be an AOT-friendly alternative, but in turn that depends on having all type usage discoverable through static analysis (not coming at runtime from external references, for example) - it's harder to figure out if that's going to be the case here :)

Reply:
Not suggesting this should block anything for preview 6, or even for .NET 8 overall - just trying to get a sense of whether this changes anything longer term.

Reply:
> Is there anything we should be concerned about regarding AOT/trimming when using MakeGenericType? I know that the rest of the server-side system likely isn't fully compatible with AOT/trimming yet so it's not like this would be a regression. Just thinking ahead to if/when we try to get full AOT/trim compat. Will this be a blocker, or does it just work, or will there be some other way to achieve what we want at that point?

> I know that source generators can in some cases be an AOT-friendly alternative, but in turn that depends on having all type usage discoverable through static analysis (not coming at runtime from external references, for example) - it's harder to figure out if that's going to be the case here :)

It'll be a problem when we move to AoT, but it's no different than any other reflection pattern in the framework. We'll cross that bridge when we are there.

Reply:
> no different than any other reflection pattern in the framework

Not really accurate. Many kinds of reflection work fine with AOT, generally those that involve reading type data. The ones that challenge AOT are runtime codegen specifically.

Reply:
Fair enough, I was trying to imply that AoT is something that we need to considerate as a whole when/if we decide to tackle it.